src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -0,0 +1,282 @@
+using MessagePack;
+using System.Buffers;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using TypeShape.Abstractions;
+using TypeShape.Examples.Utilities;
+
+namespace TypeShape.Examples.MsgPackSerializer;
+
+/// <summary>
+/// A msgpack serializer built on top of TypeShape.
+/// </summary>
+public static class MsgPackSerializer
+{
+    /// <summary>
+    /// Serializes a value.
+    /// </summary>
+    /// <param name="writer">The writer that receives the msgpack bytes.</param>
+    /// <param name="value">The value to be serialized.</param>
+    /// <typeparam name="T">The type of value to be serialized.</typeparam>
+    public static void Serialize<T>(IBufferWriter<byte> writer, in T? value)
+        where T : IShapeable<T>
+    {
+        MessagePackWriter msgpackWriter = new(writer);
+        SerializerCache<T>.Serializer(ref msgpackWriter, value);
+        msgpackWriter.Flush();
+    }
+
+    /// <summary>
+    /// Deserializes msgpack into a value of a particular type.
+    /// </summary>
+    /// <param name="sequence">The msgpack to deserialize.</param>
+    /// <typeparam name="T">The type of value to deserialize.</typeparam>
+    /// <returns>The deserialized value.</returns>
+    public static T? Deserialize<T>(ReadOnlySequence<byte> sequence)
+        where T : IShapeable<T>
+    {
+        MessagePackReader reader = new(sequence);
+        return SerializerCache<T>.Deserializer(ref reader);
+    }
+
+    private delegate T? Reader<T>(ref MessagePackReader reader);
+
+    private delegate void Writer<T>(ref MessagePackWriter writer, T value);
```

Comment:
Couldn't the two delegates be squished into a common `MessagePackConverter<T>` or `MessagePackFormatter<T>` type? Or are you intentionally looking to separate the two?

Reply:
For what it's worth, messagepack has users where they can only serialize or deserialize a type, not both. So it made sense to me to write this so that they were independently generated.

Reply:
That said, if I end up implementing this as an `IMessagePackFormatter<T>`, then yes I'll be implementing both directions as one unit.